neptune/internal/artifacts/types.py

```diff
@@ -0,0 +1,78 @@
+#
+# Copyright (c) 2021, Neptune Labs Sp. z o.o.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+import abc
+import enum
+import pathlib
+import typing
+from dataclasses import dataclass
+
+from neptune.new.exceptions import NeptuneUnhandledArtifactSchemeException, NeptuneUnhandledArtifactTypeException
+
+
+class ArtifactFileType(enum.Enum):
+    S3 = "S3"
+    LOCAL = "Local"
+
+
+@dataclass
+class ArtifactFileData:
+    file_path: str
+    file_hash: str
+    type: str
+    metadata: typing.Dict[str, str]
+
+
+class ArtifactDriversMap:
+    _implementations: typing.List['ArtifactDriver'] = []
+
+    @classmethod
+    def match_path(cls, path: str) -> 'ArtifactDriver':
+        for artifact_driver in cls._implementations:
```

Comment:
Instead of storing implementations list consider the profs/cons of `ArtifactDriver.__subclasses__()`

Reply:
1. ArtifactDriver.__subclasses__() only lists immediate subclasses (you'd have to do recursion or remember to NEVER do sub-specialization)
2. I like to store/use instances. This way typing is nicer (I don't have to annotate `typing.Type['ArtifactDriver']` everywhere), and it feels more natural to NOT use static-/classmethods everywhere ;)